service-base/src/main/java/org/eclipse/hono/service/registration/AbstractRegistrationService.java

```diff
@@ -335,4 +422,38 @@ protected JsonArray getSupportedGatewaysForDevice(final String tenantId, final S
         return viaObj instanceof JsonArray ? (JsonArray) viaObj : new JsonArray(Collections.emptyList());
     }
 
+    /**
+     * Gets the list of gateways that may act on behalf of the given device.
+     * <p>
+     *
+     * This default implementation gets the list of gateways from the value of the
+     * {@link RegistrationConstants#FIELD_VIA} property in the device's registration information.
+     * If a gatewayId is already known (e.g. through an assert request) this device is also included in the return value.
+     * If no acting gateway is known and the {@link RegistrationConstants#FIELD_VIA} is empty the groups in the
+     * {@link RegistrationConstants#FIELD_VIA_GROUPS} property in the device's registration information are resolved and
+     * returned.
+     * <p>
+     * Subclasses may override this method to provide a different means to determine the supported gateways.
+     *
+     * @param tenantId The tenant id.
+     * @param deviceId The device id.
+     * @param registrationInfo The device's registration information.
+     * @param resultHandler The handler to invoke with the result of this operation.
+     */
+    protected void getSupportedGatewaysForDevice(final String tenantId, final String deviceId, final JsonObject registrationInfo,
```

Comment:
Same as above: It is preferable to have the `Future<JsonArray>` return value here instead of the `Handler` parameter.
The other `getSupportedGatewaysForDevice` method can be removed.